src/controllers/authenticationController.ts

```diff
@@ -105,6 +108,80 @@ export default class AuthenticationController {
       .send(metadata);
   }
 
+  /**
+   * Returns data about the current user session.
+   */
+  public async getSessionState(
+    req: express.Request
+  ): Promise<Either<Error, ISessionState>> {
+    try {
+      const maybeToken = await this.extractTokenFromRequest(req);
+      if (isNone(maybeToken)) {
+        return left(new Error("No token in the request"));
+      }
+
+      const errorOrSession = await this.sessionStorage.get(maybeToken.value);
+      if (isLeft(errorOrSession)) {
+        const errorOrSessionState = await this.refreshUserToken(
+          maybeToken.value
+        );
+
+        if (isLeft(errorOrSessionState)) {
+          const error = errorOrSessionState.value;
+          return left(error);
+        }
+
+        const sessionState = errorOrSessionState.value;
+        return right(sessionState);
+      }
+
+      const session = errorOrSession.value;
+      const publicSession = {
+        expireAt: session.expireAt,
+        expired: session.expired
+      };
+      return right(publicSession);
+    } catch (err) {
+      return left(new Error("Error while loading the session"));
+    }
+  }
+
+  /**
+   * Extracts the Bearer token from the authorization header in the request.
+   */
+  private async extractTokenFromRequest(
+    req: express.Request
+  ): Promise<Option<string>> {
+    if (req.headers && req.headers.authorization) {
+      const authorization = req.headers.authorization as string;
+      const parts = authorization.split(" ");
+      if (parts.length === 2) {
+        const scheme = parts[0];
+        const token = parts[1];
+
+        if (/^Bearer$/i.test(scheme)) {
+          return some(token);
+        } else {
+          return none;
+        }
+      }
+    }
+    return none;
+  }
+
+  /**
+   * Refresh a user token in the session storage and returns the new token.
+   */
+  private async refreshUserToken(
+    token: string
+  ): Promise<Either<Error, ISessionState>> {
+    try {
+      return await this.sessionStorage.refresh(token);
+    } catch (err) {
+      return left(new Error("Error while refreshing the token"));
```

Comment:
does this ever happen? since `sessionStorage.refresh` returns an `Either`, it's supposed to never get rejected...

Reply:
I guess it could happen if this.redisClient.hgetall inside refresh throws an exeption

Reply:
I mean, we can either return untyped errors via `Promise<T>` or return typed errors with `Promise<Either<E, T>>` - imho the latter form signals the caller that the `Promise` will never be rejected and all errors will be catched and returned as `Promise<Left<E>>` - at least this is what I think we did in the CD APIs.

Mixing the two doesn't make sense.